clarity/utils/audiogram.py

```diff
@@ -0,0 +1,170 @@
+        x_in (ndarray): x-axis values, e.g. audiogram frequencies (Hz)
+        y_in (ndarray): y-axis values, e.g., audiogram levels
+        x_out (ndarray): x-axis value at which y-value are required
+
+    Returns:
+        ndarray: interpolated y-values
+
+    """
+
+    return np.interp(np.log(x_out), np.log(x_in), y_in, left=y_in[0], right=y_in[-1])
+
+
+@dataclass
+class Audiogram:
+    """Dataclass to represent an audiogram.
+
+    Attributes:
+        levels (ndarray): The audiometric levels in dB HL
+        frequencies (ndarray): The frequencies at which the levels are measured
+
+    """
+
+    levels: np.ndarray
+    frequencies: np.ndarray = np.array(DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES)
+
+    def __post_init__(self) -> None:
+        """Check that dimensions of levels and frequencies match."""
+        if len(self.levels) != len(self.frequencies):
+            raise ValueError(
+                f"Levels ({len(self.levels)}) and frequencies ({len(self.frequencies)})"
+                " must have the same length"
+            )
+
+    @property
+    def severity(self) -> str:
+        """Categorise HL severity level for the audiogram.
+
+        Note that this categorisation is different from that of the British
+        Society of Audiology, which recommends descriptors mild, moderate,
+        severe and profound for average hearing threshold levels at 250, 500,
+        1000, 2000 and 4000 Hz of 21-40 dB HL, 41-70 dB HL, 71-95 dB HL
+        and > 95 dB HL, respectively (BSA Pure-tone air-conduction and
+        bone-conduction threshold audiometry with and without masking
+        2018).
+
+        Returns:
+            str -- severity level, one of SEVERE, MODERATE, MILD, NOTHING
+
+        """
+        # calculate mean hearing loss between critical frequencies of 2 & 8 kHz
+        critical_freqs = np.logical_and(
+            2000 <= self.frequencies, self.frequencies <= 8000
+        )
+        critical_levels = self.levels[critical_freqs]
+        # Remove any None values
+        critical_levels = [x for x in critical_levels if x is not None]
+        # Ignore any None values
+        impairment_degree = np.mean(critical_levels) if len(critical_levels) > 0 else 0
+
+        if impairment_degree > 56.0:
+            return "SEVERE"
+        if impairment_degree > 35.0:
+            return "MODERATE"
+        if impairment_degree > 15.0:
+            return "MILD"
+
+        return "NOTHING"
+
+    def resample(self, new_frequencies: ndarray) -> Audiogram:
+        """Resample the audiogram to a new set of frequencies.
+
+        Interpolates linearly on a log frequency axis.
+
+        Args:
+            new_frequencies (ndarray): The new frequencies to resample to
+
+        Returns:
+            Audiogram: New audiogram with resampled frequencies
+
+        """
+        return Audiogram(
+            levels=logx_interp(self.frequencies, self.levels, new_frequencies),
+            frequencies=new_frequencies,
+        )
```

Comment:
Perhaps, we could validate first that `self.frequency` and `new_frequency` are different before computing the `logx_interp`.
If they are equal, return the `self.Audiogram`

Reply:
Thanks. I'll reconsider this.

For that one case alone, I'm not sure the efficiency saving is worth the complication. The current function returns a new audiogram so you would still need to make a copy or the behaviour would be unpredictable so you would end up with something like,

```
 if np.all(new_frequencies == self.frequencies):
            return copy.deepcopy(self)
```

But elsewhere in the code though, we had a method for making an audiogram from a subset of existing audiogram frequencies which is now being done using the resample method -- but this doesn't validate that the subset existed. So perhaps the resample could have a 'validate' flag the is defaults to False but if True will only work if the new_frequencies were a subset ... so that you can guarantee that the new audiogram is made of genuine HL measurements not interpolated.